src/Http/Routing/src/DefaultEndpointConventionBuilder.cs

```diff
@@ -12,24 +12,37 @@ internal class DefaultEndpointConventionBuilder : IEndpointConventionBuilder
     {
         internal EndpointBuilder EndpointBuilder { get; }
 
-        private readonly List<Action<EndpointBuilder>> _conventions;
+        private List<Action<EndpointBuilder>>? _conventions;
 
         public DefaultEndpointConventionBuilder(EndpointBuilder endpointBuilder)
         {
             EndpointBuilder = endpointBuilder;
-            _conventions = new List<Action<EndpointBuilder>>();
+            _conventions = new();
         }
 
         public void Add(Action<EndpointBuilder> convention)
         {
-            _conventions.Add(convention);
+            var conventions = _conventions;
+
+            if (conventions is null)
+            {
+                throw new InvalidOperationException("Conventions cannot be added after building the endpoint");
+            }
+
+            conventions.Add(convention);
         }
 
         public Endpoint Build()
         {
-            foreach (var convention in _conventions)
+            // Only apply the conventions once
+            var conventions = Interlocked.Exchange(ref _conventions, null);
```

Comment:
It's interlocked because we're worried about `Add` / `Build` being called concurrently?

Reply:
I'm just being extra defensive. It's unclear if these APIs are intended